src/basic/copy.c

```diff
@@ -530,13 +531,38 @@ static int fd_copy_directory(
                         continue;
                 }
 
-                if (buf.st_dev != original_device)
-                        continue;
+                if (S_ISDIR(buf.st_mode)) {
+                        /*
+                         * Don't descent into directories on other file systems. We do a simple .st_dev check here,
+                         * which basically comes for free. Note that we do this check only on directories, not other
+                         * kind of file system objects, for two reason:
+                         *
+                         * • The kernel's overlayfs pseudo file system that overlays multiple real file systems
+                         *   propagates the .st_dev field of the file system a file originates from all the way up
+                         *   through the stack to stat(). It doesn't do that for directories however. This means that
+                         *   comparing .st_dev on non-directories suggests that they all are mount points. To avoid
+                         *   confusion we hence avoid relying on this check for regular files.
+                         *
+                         * • The main reason we do this check at all is to protect ourselves from bind mount cycles,
+                         *   where we really want to avoid descending down in all eternity. However the .st_dev check
+                         *   is usually not sufficient for this protection anyway, as bind mount cycles from the same
+                         *   file system onto itself can't be detected that way.
+                         */
+
+                        if (!FLAGS_SET(copy_flags, COPY_CROSS_MOUNTS)) {
+                                if (buf.st_dev != original_device)
+                                        continue;
+
+                                r = fd_is_mount_point(dirfd(d), de->d_name, 0);
```

Comment:
I'm a bit worried that this might become too expensive too quickly, particularly on a tree with many directories.

Also, not sure it solves 100% of the problem, since you could still find ways to create infinite trees within the same filesystem (I'm thinking FUSE.)

It also might have impact on some legitimate use cases for bind mounting the same filesystem, for instance bind mounting a directory onto itself to change its permissions and mode (e.g. shared, ro/rw, etc.)

Perhaps a simpler approach would be to just count the levels and break after a limit is reached? Considering every directory adds at least two characters to the path (the non-empty directory name + a slash), perhaps using `PATH_MAX / 2` (or perhaps something based on `pathconf(_PC_PATH_MAX)`) would be a good simple cheap approach to breaking infinite loops?